internal/hclogutils/args.go

```diff
@@ -10,27 +14,47 @@ func MapsToArgs(maps ...map[string]interface{}) []interface{} {
 		result := make([]interface{}, 0, len(maps[0])*2)
 
 		for k, v := range maps[0] {
-			result = append(result, k)
-			result = append(result, v)
+			result = append(result, k, v)
 		}
 
 		return result
 	default:
-		mergedMap := make(map[string]interface{}, 0)
+		// Pre-allocate a map to merge all the maps into,
+		// that has at least the capacity equivalent to the number
+		// of maps to merge
+		mergedMap := make(map[string]interface{}, len(maps))
 
+		// Merge all the maps into one;
+		// in case of clash, only the last key is preserved
 		for _, m := range maps {
 			for k, v := range m {
 				mergedMap[k] = v
 			}
 		}
 
-		result := make([]interface{}, 0, len(mergedMap)*2)
+		// As we have merged all maps into one, we can use this
+		// same function recursively for the `switch case 1`.
+		return MapsToArgs(mergedMap)
+	}
+}
 
-		for k, v := range mergedMap {
-			result = append(result, k)
-			result = append(result, v)
-		}
+// ArgsToKeys will extract all keys from a slice of key/value pairs
+// arguments (i.e. `[k1, v1, k2, v2, ...]`) expected by hc-log.Logger methods
+func ArgsToKeys(args []interface{}) []string {
+	// Pre-allocate enough capacity to fit all the keys,
+	// i.e. all the elements in the input array in even position
+	keys := make([]string, 0, len(args)/2)
 
-		return result
+	for i := 0; i < len(args); i += 2 {
```

Comment:
Although it looks like this should never happen, if `len(args)` is an odd/uneven number, then a key with no corresponding value could be appended to `keys`.

Perhaps `for i := 0; i + 1 < len(args); i += 2 {` could be used to avoid such a scenario?